relay-cardinality/src/redis.rs

```diff
@@ -1,27 +1,145 @@
-use std::{collections::HashSet, time::SystemTime};
+use std::time::SystemTime;
 
 use itertools::Itertools;
 use relay_redis::{
-    redis::{self, ToRedisArgs},
+    redis::{self, FromRedisValue, Script},
     Connection, RedisPool,
 };
 use relay_statsd::metric;
 
 use crate::{
     limiter::{CardinalityScope, Entry, EntryId, Limiter, Rejection},
-    statsd::CardinalityLimiterCounters,
+    statsd::{CardinalityLimiterCounters, CardinalityLimiterHistograms},
     OrganizationId, Result,
 };
 
 /// Key prefix used for Redis keys.
 const KEY_PREFIX: &str = "relay:cardinality";
-/// Maximum amount of arguments for a Redis `SADD` call.
-const MAX_SADD_ARGS: usize = 1000;
+
+struct CardinalityScript(Script);
+
+/// Status wether an entry/bucket is accepted or rejected by the cardinality limiter.
+#[derive(Clone, Copy, Debug)]
+enum Status {
+    /// Item is rejected.
+    Rejected,
+    /// Item is accepted.
+    Accepted,
+}
+
+impl Status {
+    fn is_rejected(&self) -> bool {
+        matches!(self, Self::Rejected)
+    }
+}
+
+impl FromRedisValue for Status {
+    fn from_redis_value(v: &redis::Value) -> redis::RedisResult<Self> {
+        let accepted = bool::from_redis_value(v)?;
+        Ok(if accepted {
+            Self::Accepted
+        } else {
+            Self::Rejected
+        })
+    }
+}
+
+#[derive(Debug)]
+struct CardinalityScriptResult {
+    cardinality: u64,
+    rejected: usize,
+    stati: Vec<Status>,
```

Comment:
As much as I like Latin: https://en.wiktionary.org/wiki/statuses#English

```suggestion
    statuses: Vec<Status>,
```